fix: add jupyter api runtime

opqrstuvcut opened the pull request:
I have added a feature that enables the use of kernels launched by the jupyter lab command. This is accomplished by calling the Jupyter server API.

You can verify the functionality with the following steps:
1. Start the server using the jupyter lab command (for example: jupyter lab --ip=0.0.0.0 --port=8080 --allow-root)
2. In NeoVim, run :MoltenInit http://localhost:8080

The issue mentioned below has been fixed: https://github.com/benlubas/molten-nvim/issues/215#issuecomment-2274812321

opqrstuvcut committed:
add jupyter api runtime

opqrstuvcut committed:
Merge branch 'main' into feature/add_jupyterapi_runtime

opqrstuvcut committed:
Merge branch 'main' into feature/add_jupyterapi_runtime

opqrstuvcut committed:
add timeout

opqrstuvcut committed:
Merge branch 'main' into feature/add_jupyterapi_runtime

opqrstuvcut committed:
Merge branch 'main' into feature/add_jupyterapi_runtime

opqrstuvcut committed:
fix typo

opqrstuvcut committed:
Merge branch 'main' into feature/add_jupyterapi_runtime

opqrstuvcut committed:
fix bug

opqrstuvcut committed:
add stdin_msg

opqrstuvcut committed:
Merge branch 'feature/add_jupyterapi_runtime'

opqrstuvcut committed:
Merge remote-tracking branch 'upstream/main'

opqrstuvcut committed:
fix: discard unnecessary messages at first.

opqrstuvcut committed:
Merge remote-tracking branch 'upstream/main'

opqrstuvcut committed:
Merge branch 'main' into feature/add_jupyterapi_runtime

opqrstuvcut committed:
fix: add dummy function

benlubas commented on rplugin/python3/molten/jupyter_server_api.py:
```diff
@@ -0,0 +1,134 @@
+from urllib.parse import urlparse
+
+import requests
+import websocket
```
These are both packages that need to be installed, right? Can we put them inside the class/functions that they're used in to avoid erroring if people want to use the plugin without this feature and without installing these packages?

These requirements also need to be documented in the README

benlubas commented on rplugin/python3/molten/jupyter_server_api.py:
```diff
@@ -0,0 +1,134 @@
+        parsed_url = urlparse(url)
+        self._base_url = f"{parsed_url.scheme}://{parsed_url.netloc}"
+
+        token_part = re.search(r"token=(.*)", parsed_url.query)
```
I know it's not common, but wouldn't this fail for a url that had multiple query params?

Can we get this information from `parsed_url`? Or just make this query search up to `$` or `&`

benlubas reviewed (changes requested):
A couple comments. In general this is an awesome feature to have, thank you for contributing.

Seeing the implementation of `JupyterAPIClient`, I can't help but think that surely `jupyter_client` has already implemented something like this. Have you checked? I might try to poke around the code a little bit myself.

opqrstuvcut committed:
fix: modify the method for token query parameters.

opqrstuvcut committed:
fix: lazy load requests and websockets modules as required

opqrstuvcut committed:
feat: use jupyter server api

opqrstuvcut commented:
Thank you for your review. I have fixed the issues you pointed out, so please confirm the changes.

> Seeing the implementation of JupyterAPIClient, I can't help but think that surely jupyter_client has already implemented something like this. Have you checked? I might try to poke around the code a little bit myself.

It might also be possible to achieve this with jupyter_client. At the time I implemented the core part, I wasn't sure if jupyter_client could accomplish what I wanted, which is why I believe the current implementation turned out the way it did.
